text/0813-template-layering.md

```diff
@@ -0,0 +1,857 @@
+---
+Stage: Accepted
+Start Date: 2022-04-16
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/813
+---
+
+# Layering and Desugaring for First-Class Component Templates
+
+## Summary
+
+This RFC propose to introduce a `template()` function as an alternative for the
+`<template>` syntax extension defined in [RFC #779][rfc-779] using standard
+JavaScript, for use in environments where the custom syntax extension is not
+available, and also as a publishing format for addons on NPM:
+
+[rfc-779]: ./0779-first-class-component-templates.md
+
+```js
+import { template } from "@ember/template-compilation";
+import Hello from "my-app/components/hello";
+
+// <template><Hello></template> becomes...
+export default template(`<Hello />`, () => ({ Hello }));
+
+// export const Foo = <template><Hello /></template> becomes...
+export const Foo = template(`<Hello />`, () => ({ Hello }));
+
+// export class Bar {
+//   <template><Hello /></template>
+// }
+export class Bar {
+  static {
+    template(`<Hello />`, () => ({ Hello }), this);
+  }
+}
+```
+
+Idiomatic usages of the `template()` function can be pre-compiled by a build
+step if desired, but a runtime implementation can also be used via an opt-in
+described below.
+
+We also propose an alternative form that uses the `template()` function as a
+tag function for JavaScript tagged template strings. However, this variant does
+not have a runtime implementation and must be handled by a build step:
+
+```js
+import { template } from "@ember/template-compilation";
+import Hello from "my-app/components/hello";
```

Comment:
I wonder if this import wouldn't be detected as inactive by the syntax hyghlighter :thinking:

Reply:
I pushed some updates that clarified the motivations for the RFC. This is definitely not intended to _replace_ the `<template>` proposal.

Regarding your specific question, yeah, it is likely that without a lot of additional configurations/integrations the editor will probably not be able syntax highlight the content of the code block. Variables that are referenced only from within the template blocks will likely be (incorrectly) marked as unused.

These drawbacks are enumerated in RFC #779. Perhaps counterintuitively, the experience from the prototyping phase was that it would be _more_ difficult to properly support these things inside a `.js`/`.ts` file, because the `.js`/`.ts` language support plugins generally does not provide a way to do that for (what appears to be) arbitrary string literals. That's why the `<template>`+ custom file type approach was chosen instead. We'd have to do the work of wrapping/forking the JS/TS language support plugins _anyway_, and overall we believe it will be a better experience/more practical to have these Ember-specific plugins target Ember-specific file formats.

This RFC does not suggest that we revisit any of these, but just acknowledging that there are situations in which it is not feasible/possible to use/support the custom file format, and it will still be important to have _some_ (even if degraded) form of the `<template>` feature/programming model available in those environments.